Bug 1260899 - Handle inconsistency when merging buffer bookmarks

sleroux opened the pull request:
As per the email thread chain, this PR contains two parts:

1. The ability to detect validation failures in the Bookmarks buffer and bubble those errors up to an external caller.

2. An API for clients to query the bookmarks provider allow them to ask when bookmarks can be modified and which actions they can take.

sleroux commented:
@rnewman I think I lost your previous comment when I pushed up the new changes. With regards to the comment, I agree that the old commit was lying to the caller saying that the sync operation was not started when in it fact it was - it just failed. I've updated the code to now pass along the error instead of telling the user .NotStarted.

Along with that commit I've added a small API to bookmarks that checks to see if there are contents within the buffer for MergedSQLBookmarks. This API returns a list of actions the client can take on the bookmarks. I can simplify this down to Read, Read | Write if we don't want this level of granularity.

sleroux commented:
Ended up changing it to return ReadOnly/ReadWrite states based on if the bookmarks buffer is empty or not.

rnewman commented on Storage/Bookmarks/Bookmarks.swift:
```diff
@@ -22,6 +22,15 @@ public protocol SyncableBookmarks: class, ResettableSyncStorage, AccountRemovalD
     func applyLocalOverrideCompletionOp(op: LocalOverrideCompletionOp, itemSources: ItemSources) -> Success
 }
 
+public enum BookmarksEditState {
```
This might be more accurate if you have `NoUnmergedRemoteChanges` and `UnmergedRemoteChanges` — it's the UI that's deciding that those two should be interpreted as RW/RO.

rnewman commented on Storage/SQL/SQLiteBookmarksSyncing.swift:
```diff
@@ -638,6 +638,14 @@ extension MergedSQLiteBookmarks: ShareToDestination {
 
+extension MergedSQLiteBookmarks: EditableBookmarks {
+    public func editState() -> Deferred<Maybe<BookmarksEditState>> {
+        return self.buffer.isEmpty() >>== { isEmpty in
```
I think

```
return chain(self.buffer.isEmpty()) {
     $0 ? .NoUnmergedRemoteChanges : UnmergedRemoteChanges
}
```

rnewman commented on Sync/Synchronizers/Bookmarks/BookmarksSynchronizer.swift:
```diff
@@ -178,9 +178,13 @@ class MergeApplier {
-          >>== self.applyResult
-           >>> always(SyncStatus.Completed)
+        return self.merger.merge().bind { result in
+            // Pass on any failures we get when trying to merge
```
That's what the chaining operators *do*. What am I missing about this change?

rnewman commented on Sync/Synchronizers/Bookmarks/BookmarksSynchronizer.swift:
```diff
@@ -361,12 +365,16 @@ class ThreeWayBookmarksStorageMerger: BookmarksStorageMerger {
                     // No outgoing changes. Unilaterally apply remote changes if they're consistent.
-                    return self.buffer.validate() >>> self.applyIncomingDirectlyToMirror
+                    return self.buffer.validate().bind { result in
+                        result.isSuccess ? self.applyIncomingDirectlyToMirror() : deferMaybe(result.failureValue ?? BookmarksMergeConsistencyError())
```
Again, if validation fails, it'll fill the deferred with a failure, and won't move on to the next step. Why do you need `bind`?

rnewman commented on Sync/Synchronizers/Bookmarks/BookmarksSynchronizer.swift:
```diff
@@ -361,12 +365,16 @@ class ThreeWayBookmarksStorageMerger: BookmarksStorageMerger {
                     // Changes on both sides. Merge.
-                    return self.buffer.validate() >>> self.threeWayMerge
+                    return self.buffer.validate().bind { result in
+                        result.isSuccess ? self.threeWayMerge() : deferMaybe(result.failureValue ?? BookmarksMergeConsistencyError())
```
Same.

rnewman commented on SyncTests/TestBookmarkTreeMerging.swift:
```diff
@@ -290,6 +290,114 @@ class TestBookmarkTreeMerging: FailFastTestCase {
+        let storer = uploader.getStorer()
+        let applier = MergeApplier(buffer: bookmarks, storage: bookmarks, client: storer, greenLight: { true })
+        let result = applier.go().value
+        XCTAssertFalse(result.isSuccess)
```
We have a test helper:

```
extension Deferred where T: Succeedable {
    func succeeded() {
        XCTAssertTrue(self.value.isSuccess)
    }
}
```

Extend that to have `.failed`, too, and use it here and throughout.

rnewman commented on SyncTests/TestBookmarkTreeMerging.swift:
```diff
@@ -290,6 +290,114 @@ class TestBookmarkTreeMerging: FailFastTestCase {
+        let result = applier.go().value
+        XCTAssertFalse(result.isSuccess)
+        XCTAssertTrue(result.isFailure)
+        XCTAssertEqual(result.failureValue!.description, BufferInconsistency.MissingStructure.description)
```
Also add tests before and after the `go` for the edit state (or whatever you rename it to): it should return read-only both times.

sleroux commented:
After our chat today I went digging into how the current bookmarks implementation determines what is editable or not. From what I can tell, when fetching folders/children, we inform it of the direction that the request is coming from - either `.Local` or `.Buffer`. Currently, we flat out say that all items from the buffer are not editable [1]. I assume that this was because we didn't support bidirectional bookmark sync and were read-only because of it. Now with bidirectional bookmark sync enabled we'll want to make this smarter and only disable deletion if we have unmerged remote bookmarks in our buffer. I don't see using the API I exposed here though. Instead of using the `isEditable` value from Swift, I was thinking of amending the getChildren query to include the result of the buffer's isEmpty query as the isEditable value. This way the editable value is always true for `.Local` and determined by the contents of the buffer for buffered bookmarks. Am I way off base here or does this make sense? @rnewman 

[1] https://github.com/mozilla/firefox-ios/blob/master/Storage/SQL/SQLiteBookmarksModel.swift#L303

rnewman commented:
The direction flag shows whether we're fetching to display (local + mirror)
— i.e., normal successful operation — or the read-only fallback view of the
buffer that we use when we haven't downloaded enough data to sync.

When we enable bidirectional sync, we will normally be using .Local here.

The change you'll need to make, I think, is to *also* check to see whether
we're fully synced: that is, whether the buffer is empty. If it is, and
we're .Local, then we're editable. Otherwise (if we're in a read-only
.Buffer view *or* the buffer is non-empty) then we're read-only.

The only time we want to allow editing of bookmarks with a non-empty buffer
table is if the mirror is empty (i.e., we're unsynced) and the item we're
trying to edit is a new local bookmark.

So our test cases:

Empty buffer: all items editable.
Non-empty buffer:
  Empty mirror (hasn't synced ever)? Local items are editable.
  Non-empty mirror (has synced ever)? Nothing is editable.

Does that make sense?

sleroux commented:
> Does that make sense?

Yup this helps a lot. I'll throw up a patch with these changes. Thanks!

sleroux commented:
@rnewman I've added some logic to perform the checks you mentioned before fetching the bookmarks from the DB. I've also added some unit tests to verify. I believe this should handle all states that you mentioned. 

An interesting thing I found when writing the tests is that although we discard and create a new database each time using the `getSyncableBookmarks` method, we actually still retain some state across tests due to the `modelFactory` being a singleton [1]. This was causing previously added bookmarks to appear in later test cases. Along with feedback for the approach to add in the editable checks, does it still make sense to keep the `modelFactory` a singleton?

[1] https://github.com/mozilla/firefox-ios/pull/1780/commits/95e06d1a7cae9404b4589951c2a32e31a080b40f#diff-501066006fdca7138dbe656b8a79d6b6R33

rnewman commented:
@sleroux the model factory should be a singleton in that there should be one per `SQLiteBookmarksBase` instance. It should be relatively easy to make that a (safe, which lazy vars aren't) per-instance lazy var, no?

rnewman commented on Storage/Bookmarks/Bookmarks.swift:
```diff
@@ -22,6 +22,15 @@ public protocol SyncableBookmarks: class, ResettableSyncStorage, AccountRemovalD
+}
+
+public protocol BufferedBookmarks: class {
+    func bufferState() -> Deferred<Maybe<BookmarkBufferState>>
```
I don't see you call this anywhere in this PR. Am I missing something?

Even so, I question whether this needs to be a separate protocol. This should be hidden within the factory source, no?

sleroux replied:
Actually with the changes in the queries below, this isn't needed. I'll remove.

rnewman commented on Storage/SQL/SQLiteBookmarksModel.swift:
```diff
@@ -280,7 +275,27 @@ extension SQLiteBookmarks {
 
-        return self.db.runQuery(withIcon, args: args, factory: BookmarkFactory.factory)
+        let sql = includeIcon ? withIcon : values + " ORDER BY title ASC"
+        return isEditableForDirection(direction) >>== { editable in
```
This runs *three* SQL queries in order to get the result. You only need one.

sleroux replied:
My SQL-foo is weak 😞 I'll revise the query with your suggestions below.

rnewman commented on Storage/SQL/SQLiteBookmarksModel.swift:
```diff
@@ -280,7 +275,27 @@ extension SQLiteBookmarks {
+        return accumulate([isBufferEmpty, isMirrorEmpty]) >>== { checks in
+            let bufferEmpty = checks[0]
+            let mirrorEmpty = checks[1]
+            return deferMaybe(bufferEmpty && direction == .Local ? true : mirrorEmpty)
```
If `direction != .Local`, you don't even need to run the first query, right?

But even so, here you only need one query!

```
SELECT
ifnull((SELECT 1 FROM \(TableBookmarksBuffer)), 0) AS hasBuffer,
ifnull((SELECT 1 FROM \(TableBookmarksMirror)), 0) AS hasMirror
```

gives you two columns, one for each table.

rnewman replied:
Even better, use `exists`:

```
SELECT exists(SELECT 1 FROM \(TableBookmarksBuffer)) AS hasBuffer, …
```

which *might* be smart enough that you can insert it into your `isEditable` clause above, and skip this whole function.

rnewman commented on SyncTests/TestBookmarkModel.swift:
```diff
@@ -0,0 +1,183 @@
+        let bookmarkURL = "http://AAA.com".asURL!
+        bookmarks.local.insertBookmark(bookmarkURL, title: "AAA", favicon: nil, intoFolder: BookmarkRoots.MenuFolderGUID, withTitle: "").succeeded()
+
+        XCTAssertTrue(bookmarks.isMirrorEmpty())
```
You're missing a `.value()` here to resolve the deferred. Look for this error throughout.

sleroux replied:
The `isMirrorEmpty` resolves the deferred internally but I'll expose it for better readability.

sleroux commented:
@rnewman I've updated the PR to address the feedback. Now,

1. `isEditable` is a query that resolves to a scalar value within the existing query - eliminating the other two queries we were making before.
2. Removed the unused protocol and methods for checking if the buffer contains contents or not since it has been supplanted by the changes to `getRecordsWithGUIDs` and `getChildrenWithParent`.

rnewman commented on ClientTests/MockProfile.swift:
```diff
@@ -111,7 +111,7 @@ public class MockProfile: Profile {
     }()
 
-    lazy var bookmarks: protocol<BookmarksModelFactorySource, SyncableBookmarks, LocalItemSource, MirrorItemSource, ShareToDestination> = {
+    lazy var bookmarks: protocol<BookmarksModelFactorySource, SyncableBookmarks, LocalItemSource, BufferedBookmarks, MirrorItemSource, ShareToDestination> = {
```
Don't forget to run tests, then fix this when they fail :)

sleroux replied:
Oops! I've been running the tests from a new SyncTests target and forgot to run the client target :(





On Sat, May 28, 2016 at 8:58 AM -0700, "Richard Newman" <notifications@github.com> wrote:












In ClientTests/MockProfile.swift:
> @@ -111,7 +111,7 @@ public class MockProfile: Profile {
>          return CertStore()
>      }()
>  
> -    lazy var bookmarks: protocol<BookmarksModelFactorySource, SyncableBookmarks, LocalItemSource, MirrorItemSource, ShareToDestination> = {
> +    lazy var bookmarks: protocol<BookmarksModelFactorySource, SyncableBookmarks, LocalItemSource, BufferedBookmarks, MirrorItemSource, ShareToDestination> = {



Don't forget to run tests, then fix this when they fail :)



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

rnewman commented on Storage/SQL/SQLiteBookmarksModel.swift:
```diff
@@ -252,20 +252,23 @@ public class SQLiteBookmarksModelFactory: BookmarksModelFactory {
 }
 
 extension SQLiteBookmarks {
+    private static var isEditableQuery: String {
```
Just put this at the top level as a `private let` two lines above, rather than jumping through the hoops to put it inside the extension.

rnewman commented:
If it works, ship it!

sleroux commented:
Thanks for all the feedback!
